### PR TITLE
Make by-name types first-class value types

### PR DESCRIPTION
**Pull request by odersky**
A big revision so that types indicating call-by-name passing are regular value types. In fact, the
by-name type `=> T` now behaves like a nullary context function type `() ?=> T`. Nullary context function types still cannot be written explicitly but the type mechanics for cbn types and nullary context function types are the same.

Benefits:

 - Overall, a significant simplification. The previous transformation pipeline had to do a complicated danse to expand call-by-name types. This is no longer necessary.
 - Call-by-name types are now regular value types. This removes some ad-hoc restrictions, such as that cbn methods cannot be eta expanded to function values.
 - Delayed call-by name arguments are marked explicitly from Typer. This means follow-on analyses such as initialization checking can be simplified. 
 - Potential to drop more restrictions, in particular
     - By-name types could be allowed to appear everywhere, not just in parameter position
     - Classes could have by-name value parameters
    These generalizations should be considered for future PRs.
 - Opens the way to refine by-name types to be pure or impure, or to let them have specific capture capabilities.

Things to do:

 - [ ] Update scaladoc to handle the new by-name types (scaladoc build fails currently, as do two scaladoc tests in the `scala3-bootstrapped/test` task.
 - [ ] Adapt macro expansion and reflect.quotes to the new scheme. Currently 6 tests in BootstrappedOnlyCompilationTests fail.
 - [ ] Investigate why sbt-test/analyzer-plugin fails.

### CODE REVIEW

**Comment by som-snytt** [comment]
This makes sense from the previous PR.

It's a privilege, in the era of open source, to follow along at home. Thanks!

I still have to pause a moment to distinguish nullary and nilary, although I think I helped coin the pun. It's possible I am aging faster than Odersky.

**Comment by odersky** [comment]
@smarter To make by-name types context functions, we need to overcome a tricky interaction between type variable avoidance and implicit search divergence checking. The problem is that 
we introduce empty parameter closures for by-name args in Typer, which means that any nested variables will get higher levels than than the environment. I can see that in some changed error messages where type variables now get (avoid) or (lower) prefixes. That's not nice in error messages, but there is a much bigger problem:

We can now generate new type variables for every level of by-name implicit search. This violates the basic assumption underlying implicit divergence checking that the set of type variables is fixed. Consequently, we get infinite recursion on neg/byname-implicits-11 and neg/by-name-implicits-16. Previously, this problem did not manifest itself, since we did not generate synthesized terms with nested scopes. But now we do. (I guess even before this could have arisen if we 
generated terms for context function types, but the interaction with divergence checking was never studied). 

I see two possible routes for a solution

  - Don't bump the level for nullary closures - since they do not introduce anything of interest that needs to be avoided. That will fix the problem for the new cbn encoding, and might be a good idea for performance anyway, but we would still be vulnerable for implicit search of context function types. 

 - For the purpose implicit divergence checking treat the original variables and their lifted versions as being the same. Is there a way to map a lifted variable back to its original?

Right now I am blocked by this, so if you have some advice that would be great!

**Comment by smarter** [comment]
> Is there a way to map a lifted variable back to its original?

Yes we use the same trick than depTypeVar and can map back using representedParamRef

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 1]
```diff
@@ -726,6 +726,14 @@ trait TypedTreeInfo extends TreeInfo[Type] { self: Trees.Instance[Type] =>
     case _ => tree
   }
 
+  /** An anonyous function and a closure node referring to it in a block, without any wrappigs */
```
Comment (ritschwumm) [inline compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:729]: typo: 
anonyous -> anonymous
wrappigs -> wrappings

File: compiler/src/dotty/tools/dotc/typer/Nullables.scala [hunk 2]
```diff
@@ -498,11 +498,20 @@ object Nullables:
...
    *  flow assumptions about mutable variables and suggest that it is enclosed
    *  in a `byName(...)` call instead.
+   *
+   *  This step is needed because arguments might be evaluated in calls to overloaded
+   *  methods that have both cbn and cbv parameters, so we do not know yet whether an
+   *  argument is cbn. We always typecheck such arguments under the assumption
+   *  that they are cbv, and we have to correct this assumption if the actual resolved
+   *  method is cbn. If the call is non-overloaded, we do the right thing from the start.
+   *  Inserting a byName call just makes clear that the argumnent is cbn. There is no
+   *  special treatemnt in the compiler associated with that method; it is just the
```
Comment (ritschwumm) [inline compiler/src/dotty/tools/dotc/typer/Nullables.scala:508]: typo: treatemnt -> treatment

**Comment by odersky** [comment]
@smarter I solved the problem by delaying the closure generation. In typer a call-by-name argument is represented as `<by-name>(...)` where `<by-name>` is a synthetic function in <special-ops>, analogous to throw. These applications are then replaced with closures in phase `ByNameLambda`.

**Comment by nicolasstucki** [comment]
@odersky I added some patches in https://github.com/dotty-staging/dotty/commit/36f45987968c98a843208b6fb89168e7ce14cda2. You can cherry-pick that commit. It fixes most of the macro related issues.

**Comment by odersky** [comment]
I made the following changes

 - by-name arguments are now wrapped in a new kind of tree node of class `ByName` instead of being represented as applications of a magic function.
 - `ByName` nodes are not pickled. Instead they are reconstituted on unpickling by looking at the formal parameter types of applications.

This means that no new Tasty version is needed. The pickled Tasty stays the same.

**Comment by bishabosha** [comment]
> * `ByName` nodes are not pickled. Instead they are reconstituted on unpickling by looking at the formal parameter types of applications.
> 
> This means that no new Tasty version is needed. The pickled Tasty stays the same.

@sjrd  Should it be considered an incompatibility that `def foo(arg: => Int) = { arg; arg }` in tasty now looks like
```scala
def foo(arg: => Int) = { arg.apply(); arg.apply() }
```
i.e. we now have `SELECTin(arg.type, apply [@apply (): java.lang.Object])` rather than just `arg.type`

**Comment by sjrd** [comment]
> > * `ByName` nodes are not pickled. Instead they are reconstituted on unpickling by looking at the formal parameter types of applications.
> > 
> > This means that no new Tasty version is needed. The pickled Tasty stays the same.
> 
> @sjrd Should it be considered a Tasty incompatibility that `def foo(arg: => Int) = { arg; arg }` in tasty now looks like
> 
> ```scala
> def foo(arg: => Int) = { arg.apply(); arg.apply() }
> ```
> 
> i.e. we now have `Apply(SELECTin(arg.type, apply [@apply (): java.lang.Object]), Nil)` rather than just `arg.type`

Writing this differently is not issue, *as long as* reading the old stuff still works (backward compat) and that an compiler can still read the new stuff (forward compat).

So, if the new compiler sees only `arg.type`, will it work? If yes, it is backward compatible. If not, it is not.

If the old compiler sees the `Apply(SELECTin(...), Nil)`, will it work? If yes, it is forward compatible. Otherwise, it is not.

**Comment by smarter** [comment]
> According to the PR message we would like to represent them as () ?=> A, right?

That's the desugared form, but the form users should write and we should use for pretty-printing is still `=> A`.

**Comment by odersky** [comment]
@pikinier20 No, by-name parameter types should still be rendered as `=> T`. The analogy to `() ?=> T` is just for the compiler's internal purposes.

**Comment by pikinier20** [comment]
Okey, good that we explained this. I'm going to fix it tomorrow if it's ok.

**Comment by odersky** [comment]
@nicolasstucki I integrated your changes. Now there's 5 tests that are still failing.
```
Test Report
================================================================================

9 suites passed, 2 failed, 11 total
    tests/run-macros/i5119 failed
    tests/run-macros/tasty-interpolation-1 failed
    tests/run-macros/xml-interpolation-2 failed
    tests/run-macros/tasty-unsafe-let failed
    tests/pos-macros/null-by-name failed
```

**Comment by odersky** [comment]
The Tasty interop problems look daunting, unfortunately. 

 - The new compiler can probably be made to work with old `Expr` representations, if we generalize a bit. It will be hard to test this extensively, though. But nevertheless, we can hope for backwards compatibility.
 - Old compilers cannot work with new context function representations. That would require a major retrofit of old compiler codebases.

So this means we need a mode where we still generate the old `Expr` based representations under a release flag. Which would mean that any gain in simplifications would be lost and the new implementation is strictly more complex than the old.

Maybe we should put this work on hold a bit and decide afterwards whether it's worth continuing. /cc @nicolasstucki @pikinier20.

**Comment by smarter** [comment]
> Old compilers cannot work with new context function representations. 

Is this because of the extra `.apply()` when using a by-name? Could we strip the .apply at pickling time and re-introduce them when unpickling?

>  It will be hard to test this extensively, though.

This is made much easier by  https://github.com/lampepfl/dotty/pull/14156 which added support for multi-compiler tests to our test infrastructure, from http://dotty.epfl.ch/docs/contributing/testing.html:

> There are also other suffixes indicating how some particular files are compiled:
>    _c${compilerVersion} - compile a file with a specific version of the compiler instead of the one developed on the current branch (e.g. Foo_c3.0.2.scala)
    _r${release} - compile a file with a given value of -Yscala-release flag (e.g. Foo_r3.0.scala)
>
>Different suffixes can be mixed together (their order is not important although consistency is advised), e.g. Foo_1_r3.0, Bar_2_c3.0.2.

**Comment by odersky** [comment]
> Is this because of the extra .apply() when using a by-name? Could we strip the .apply at pickling time and re-introduce them when unpickling?

It's not only that. Also, that when the expected type is a `() ?=> T` the new compiler would generate a `ByName` tree, but the old compiler would generate a closure. Generating closures
means we subtly change avoidance patterns and some tests start to fail (notably recursive implicit searches).

**Comment by odersky** [comment]
Withdrawn in favor of #14295. The latter is more conservative, and less principled, but much simpler to implement, and it causes not inter-version headaches.

**Pull request closed without merging**